crates/neon/src/sys/mod.rs

```diff
@@ -1,38 +1,106 @@
-use std::mem::MaybeUninit;
+//! Raw bindings to [Node-API][node-api]
+//!
+//! [Node-API][node-api] is Node.js's API for building native addons. Neon is
+//! predominantly a safe wrapper for Node-API and most users should prefer the
+//! the high level abstractions [outside](crate) of the sys module.
+//!
+//! However, directly using Node-API can be a useful tool for accessing low
+//! level functionality not exposed by Neon or experimenting with extensions
+//! to Neon without needing to fork the project.
+//!
+//! [node-api]: https://nodejs.org/api/n-api.html
+//!
+//! ## Initialization
+//!
+//! Before any Node-API functions may be used, [`setup`](setup) must be called at
+//! least once.
+//!
+//! ```rust,no_run
+//! # #[cfg(feature = "sys")]
+//! # {
+//! # let env = std::ptr::null_mut().cast();
+//! unsafe { neon::sys::setup(env); }
+//! # }
+//! ```
+//! **Note**: It is unnecessary to call [`setup`](setup) if
+//! [`#[neon::main]`](crate::main) is used to initialize the addon.
+//!
+//! ## Safety
+//!
+//! The following are guidelines for ensuring safe usage of Node-API in Neon
+//! but, do not represent a comprehensive set of safety rules. If possible,
+//! users should avoid calling Neon methods or holding references to structures
+//! created by Neon when calling Node-API functions directly.
+//!
+//! ### Env
+//!
+//! Neon ensures safety by carefully restricting access to [`Env`](bindings::Env)
+//! by wrapping it in a [`Context`](crate::context::Context). Usages of `Env`
+//! should follow Neon's borrowing rules of `Context`.
+//!
+//! It is unsound to use an `Env` if Rust's borrowing rules would prevent usage
+//! of the in scope `Context`.
+//!
+//! ### Values
+//!
+//! Neon [value types](crate::types) encapsulate references to
+//! [JavaScript values](bindings::Value) with a _known type_. It is unsound to
+//! construct a Neon value with a [`Value`](bindings::Value) of the incorrect type.
+//!
+//! ## Example
+//!
+//! ```rust,no_run
+//! # #[cfg(feature = "sys")]
+//! # {
+//! # let env = std::ptr::null_mut().cast();
+//! use neon::{context::SysContext, prelude::*, sys::bindings};
+//!
+//! let cx = unsafe {
+//!     neon::sys::setup(env);
+//!
+//!     SysContext::from_raw(env)
+//! };
+//!
+//! let raw_string: bindings::Value = cx.string("Hello, World!").to_raw();
+//! let js_string = unsafe { JsString::from_raw(&cx, raw_string) };
+//! # }
+//! ```
+use std::{mem::MaybeUninit, sync::Once};
 
+// XXX: Minimize changes due to structure changes
```

Comment:
Is `XXX` a TODO? IMO it's better hygiene to file TODOs as followup issues rather than leaving them in the code.